src/ui/src/core/navigation.ts

```diff
@@ -0,0 +1,70 @@
+type ParsedHash = {
+	pageKey?: string;
+	routeVars: Map<string, string>; // Stored as Map to avoid injection e.g. prototype pollution
+};
+
+const hashRegex = /^((?<pageKey>[^/]*))?(\/(?<routeVars>.*))?$/;
+const routeVarRegex = /^(?<key>[^=]+)=(?<value>.*)$/;
+
+export function getParsedHash(): ParsedHash {
+	const docHash = document.location.hash.substring(1);
+	const hashMatchGroups = docHash.match(hashRegex)?.groups;
+	const routeVars: Map<string, string> = new Map();
+	const pageKey = hashMatchGroups?.pageKey
+		? decodeURIComponent(hashMatchGroups.pageKey)
+		: undefined;
+
+	if (!hashMatchGroups) return { pageKey, routeVars };
+
+	const routeVarsSegments = hashMatchGroups.routeVars?.split("&") ?? [];
+	routeVarsSegments.forEach((routeVarSegment) => {
+		const matchGroups = routeVarSegment.match(routeVarRegex)?.groups;
+		if (!matchGroups) return;
+		const { key, value } = matchGroups;
+		const decodedKey = decodeURIComponent(key);
+		const decodedValue = decodeURIComponent(value);
+		routeVars.set(decodedKey, decodedValue);
+	});
+
+	return { pageKey, routeVars };
+}
+
+function setHash(parsedHash: ParsedHash) {
+	const { pageKey, routeVars } = parsedHash;
+
+	let hash = "";
+	if (pageKey) {
+		hash += `${encodeURIComponent(pageKey)}`;
+	}
+
+	if (routeVars.keys.length > 0) {
+		hash += "/";
+		hash += Array.from(routeVars.entries())
+			.map(([key, value]) => {
+				// Vars set to null are excluded from the hash
+
+				if (value === null) return null;
+				return `${encodeURIComponent(key)}=${encodeURIComponent(
+					value,
+				)}`;
+			})
+			.filter((segment) => segment)
+			.join("&");
```

Comment:
I think you should use `URLSearchParams` instead
```suggestion
		hash += Array.from(routeVars.entries())
			.reduce((params, [key, value]) => {
				// Vars set to null are excluded from the hash
				if (value !== null) params.set(key, value);
				return params;
			}, new URLSearchParams)
			.toString();
```

Reply:
This is a great suggestion and I recognize it's a better alternative, however I'd rather not focus on it right now as it's not Workflows (just on the commit because of some refactoring). I've created a Jira ticket with this for you to take care of later. Some tests for this would be great too.